### DISCUSSION TITLE
Should `StageError::Internal` be fatal?

### DISCUSSION THREAD

**Discussion by onbjerg** [paradigmxyz/reth#457] [Technical discussions]
Per https://github.com/paradigmxyz/reth/pull/453 there are two "categories" for the variants of `StageError`: fatal and non-fatal. 

Fatal errors stop the pipeline completely because they are assumed to be irrecoverable (e.g. db corruption), and non-fatal errors result in the pipeline retrying the stage again.

Currently, the only fatal variants in `StageError` are database related. There is one catch-all variant called `StageError::Internal` and it is unclear if this should be fatal or not: stages may want custom errors for informational purposes that are not fatal, but some errors from e.g. the standard library might be fatal (currently I don't think any stage emit errors from the standard library).

**Comment by gakonst** [comment]
Let's split the`Internal` error in 2 variants `Fatal` and `Recoverable` which box any kind of error so that the pipeline can know whether to stop or retry.

**Reply by onbjerg** [reply]
Done here: https://github.com/paradigmxyz/reth/pull/453/commits/9a66b07d81a159de1d6edcd817f7f49d2193d3f8
